### ISSUE TITLE
Feature request: enable reading of mzML from stdin in OpenSwathWorkflow

### ISSUE THREAD

**Issue by Ozzard** [OpenMS/OpenMS#4177]
Often, modern machines are deficient in disk throughput: many cores, high memory, but relatively slow spinning rust or network disks.  mzML is a horribly verbose interchange format, and generating mzML from a vendor format only to consume it once in OpenSwathWorkflow can be a waste of many minutes per file simply to write to disk and then read from disk.  An obvious way of resolving this (especially on clusters managed by central IT services where requesting a RAM disk on each of several hundred nodes isn't an option) is to allow piping from a conversion tool that generates mzML directly into OpenSwathWorkflow.

I suggest making two changes to the OpenSwathWorkflow options:
* Add - as an input filename to denote reading from stdin;
* Add an option to specify the format of the input file, rather than taking it from the file suffix.

I'd be happy to have a go at adding this capability if some kind soul could point me at the approximate area to look at; I've not managed to find the source for any of the executables so far.

**Comment by cbielow** [comment]
here are some pointers to our mzML reading/writing code:

 - all TOPP tools are at https://github.com/OpenMS/OpenMS/tree/develop/src/topp
 - mzML read/write via Xerces-C SAX parser at: https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/FORMAT/HANDLERS/MzMLHandler.cpp (with `void MzMLHandler::startElement(...)` etc

Now, it seems Xerces can be convinced to read from stdin https://xerces.apache.org/xerces-c/apiDocs-3/classStdInInputSource.html, but some code modification will be required. No big deal I'd hope...
Does ProteoWizard (for example) support output to stdout?

However, an even better alternative is to read vendor format directly (for Thermo raw files this should even be feasible... there is a ticket somewhere...; however this is not applicable to all vendors; supporting them all is like implementing and maintaining our own msconvert...).

The general idea seems very appealing, but I'd expect the time savings are not as high as one would imagine. Usually, the bottleneck is not necessarily the disk speed, but the creation of the internal data. E.g. base64 decoding and string->numeric conversion in general take a lot of time. We can surely improve these as well in our current code (e.g. SIMDify base64 decoding), but I'd recommend benchmarking before implementing the pipe option to see where most of the benefits are to be gained.

**Comment by Ozzard** [comment]
Thanks for the rapid response, Chris.  I've fought Xerces before - should be fine.  We're using the Dockerised SCIEX converter, which does support output to stdout.

I agree with your view that supporting all vendors is like maintaining your own msconvert; hence this approach.  I'm an old UNIX hacker with *ahem* years of experience, and love their architecture of small composable tools that do one job well.  Let's create something that allows the composition.

We've done limited benchmarking on a .Net implementation of a reader, which I would expect to have not-dissimilar characteristics to your approach except that we go multithreaded on the base64 decodes and spectrum creation in order to maximise read speed.  I suspect the TOPP tools don't do that, though I've not yet gone that deep into the source.  We'll get some speedup on read, but the bigger benefit is not having to write the intermediate file at all - the conversion is completely IO-bound.

The other huge benefit for us is not having to maintain intermediate files in scratch storage, potentially blowing our (shared) scratch or (shared) /tmp on the compute nodes in the cluster.  When one's got a queue of hundreds or thousands of injections* to analyse, that space saving is very significant.

*I'm the tech architect at the Stoller Biomarker Discovery Centre; this is not a theoretical situation for us!

**Comment by cbielow** [comment]
as a prelim test:
loading a 1.8 GB mzML from disk (ssd in this case) takes 27 seconds, i.e. on average 66 mb/sec.
loading the same file a second time, triggers NO disk access whatsoever (Win10, but probably identical across all modern OS), but still takes 27 seconds. Our parsing is thus the limiting factor (66 mb/sec). Now, you can tweak that here and there, and it depends on OS and CPU specs but that seems like something to urgently look at as well.

**Comment by cbielow** [comment]
PS: our mzML parser supports threading, so you can improve these numbers, but its far from linear (5 threads reduce time from 27 secs to 23 secs...)

**Comment by Ozzard** [comment]
Very useful benchmarks - thanks!  In that case, yes, we'll be completely IO-bound due to our cluster architecture.  Our mzML files are also routinely around the 25 Gbyte size, so scale appropriately :-).

**Comment by hroest** [comment]
@Ozzard I have been similarly frustrated with mzML. I have written a paper about the read / parsing speed of OpenMS: https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0125108 which has some interesting insights:

* OpenMS is on par with other C++ implementations like proteowizard, I dont think that there is much to gain
* for multi-threaded access it matters whether you do serial reading or parallel reading (using indexed mzML)
* changing the file format from mzML to another format (the OpenSwath cached file format) can achieve over 200x performance in read speed (Figure 2b) with 250 M peaks / s read performance 
* XML-parsing is the limiting factor, in a multithreaded context allowing up to 500 MB/s but the cached file formats boosts this to 3000 MB / s (see supplement)

Also, consider that when running OpenSwathWorkflow, we often see that the parsing of the XML file with multiple threads is a rather small part of the whole execution while the scoring and other computations take most of the time. I agree with Chris that the slowest part is actually the decoding of the Base64 strings (or zlib compression) which is still a bottleneck when you would pipe directly from stdin. In terms of space savings: if you are using the "cached" or "cachedWorkingInMemory" approach in OpenSwathWorkflow, then you are not saving much space since you have to make space for these cached files as well.

I think your best bet to improve speed is to 

- either read .wiff files directly
- use another, non-XML based file format such as one of the HDF5 file formats or the "cached" file format

**Comment by Ozzard** [comment]
Thanks, Hannes - I was aware of that paper, and I like the approaches described in it.

The gotcha I have as an architect is that I don't have a free choice of host system.  In particular, I don't have the choice of a system with local SSD or RAM disk.  Instead, our cluster nodes appear to have local spinning rust for /tmp (for example the WD2003FYYS 2Tbyte 7200rpm enterprise drive) or a network LUSTRE setup across 40Gbit Infiniband.  This gives us a problem: low bandwidth and high latency to the network disk that has reasonable random IOPS, or high bandwidth and high latency to the local disk that has dreadful random IOPS.  Sequential reads are OK, random reads are really quite unpleasant.  So I'm not optimising OpenSWATH speed assuming it's always going to be the bottleneck; instead, I'm optimising overall system speed given multiple constraints, which is a rather different job.  For example, there's a strong chance that I'll get better overall speed by forcing OpenSWATH to read an entire file sequentially rather than randomly if the random job is going to read more than (say) half the file.

Not a great explanation - sorry - but does it give a flavour of some of the trade-offs I'm facing?

**Comment by hroest** [comment]
> Not a great explanation - sorry - but does it give a flavour of some of the trade-offs I'm facing?

yes and we have the same tradeoffs in our cluster system. What we do is use the "cachedWorkingInMemory" option of OpenSwathWorkflow that writes N +1 files to the local disk where N= number of swaths. These files are then only accessed sequentially once and loaded into memory, the computation is performed on them and then they are not required any more. Can you describe in a bit more detail what your current workflow is, where your bottlenecks are and what you would like to achieve? Maybe we have already encountered a similar issue in the past...

PS: if your bottleneck is disk space, as you seem to imply above, I suggest to use the numpress compression for compressing mzML files, we get smaller files than raw .wiff , see https://github.com/ms-numpress/ms-numpress/pull/9

**Comment by Ozzard** [comment]
I'll experiment with cachedWorkingInMemory and get back to you - thanks!  I'll also look at what tooling we're using and whether it supports numpressed output from the wiff-to-mzML conversion.

**Comment by hroest** [comment]
I suggest to use the example from the documentation: http://openswath.org/en/latest/docs/openswath.html#example which contains the correct options for this to work, make sure you also set `-tempDirectory` to your *local* scratch temp directory and not the LUSTRE directory.  These are the "cached" files that I describe in the publication and they are really, really fast for IO.

**Comment by hroest** [comment]
>  I'll also look at what tooling we're using and whether it supports numpressed output from the wiff-to-mzML conversion.

in case you are interested to save space, you can use the FileFilter tool on one of your mzML file for *lossy* compression:


```
FileFilter -in test.mzML -out compressed.mzML -peak_options:numpress:masstime linear -peak_options:numpress:lossy_mass_accuracy 1e-4 -peak_options:numpress:intensity slof -peak_options:numpress:float_da slof -peak_options:zlib_compression  true 
```

since this is lossy, make sure that the results are not changed and in case you want more / less accuracy, use the `-peak_options:numpress:lossy_mass_accuracy` parameter

**Comment by hroest** [comment]
@Ozzard does this issue still affect you?

**Comment by Ozzard** [comment]
> @Ozzard does this issue still affect you?

Acutely, no; we can work round it, it's just somewhat painful.  Chronically, yes; at some point, it would still be lovely to have a workaround and be free of the nagging pain!
